### PR TITLE
Move to async generator functions for paged results

### LINKED ISSUE #706: Use PEP 525 async generator to improve data and orders client design
[PEP525](https://peps.python.org/pep-0525/) added async generators to Python 3.6. An async generator is an `async def` function or method that uses `yield` instead of `return`. Using it for SDK methods that have multiple results gives us a cleaner design that is in harmony with the designs of async Python software like aiohttp and httpx.

Here is a short bit of code that uses the data client and orders client together. We search the catalog for items using DataClient and then use a list of item ids in creating an order with OrdersClient.

```python
async def main():
    """Use DataClient and OrdersClient to search for and order scenes.
    """
    async with Session() as sess:
        data_client = DataClient(sess)

        # A reader might be led to believe that since we're awaiting the results
        # of search(), the return value has all the results materialized in it.
        search_result_iterator = await data_client.search(["PSScene"], filter_config)

        # But it does not. We still need to asynchronously iterate.
        item_ids = [item["id"] async for item in search_result_iterator]

        client = OrdersClient(sess)
        order_request = {
            "name": "my_order",
            "products": [
                {
                    "item_ids": item_ids,
                    "item_type": "PSScene",
                    "product_bundle": "analytic_udm2",
                }
            ]
        }
        order = await client.create_order(order_request)

        # Wait for order to be completed.
        order_id = order["id"]
        state = await client.wait(order_id, state="success")
        ...
```

To reiterate what I wrote in a comment, a reader of the code might be led to believe that since we're awaiting the results of search(), the return value has all the results already materialized in it. And requiring further iteration then looks confusing. I think it looks odd to await a method call and then asynchronously iterate over the return value. And Python doesn't require it. Neither httpx's [aiter_bytes](https://www.python-httpx.org/async/#streaming-responses) nor aiohttp's [iter_chunked](https://docs.aiohttp.org/en/stable/client_quickstart.html#streaming-response-content) return chunkers. They are async generators.

As an improvement, I propose to make DataClient.search an async generator and allow more concise usage like this:

```python
async def main():
    """Use DataClient and OrderClient to search for and order scenes.
    """
    async with Session() as sess:
        data_client = DataClient(sess)
        item_ids = [item["id"] async for item in data_client.search(["PSScene"], filter_config)]
        ...
```

The new implementation would be almost trivial. It would mean, for DataClient, changing the code at 
 https://github.com/planetlabs/planet-client-python/blob/main/planet/clients/data.py#L185 from

```python
        return Items(request, self._do_request, limit=limit)
```

to

```python
        async for item in Items(request, self._do_request, limit=limit):
            yield item
```

This is not unlike what httpx does here: https://github.com/encode/httpx/blob/8152c4facd0f71e0f376287e41a0810a60fec9c6/httpx/_models.py#L921. It *does* have a chunker, but it's used internally, not by httpx users.

What makes the implementation not trivial is updating tests. That's where most of the work is. I think it's worth doing. 

To support less common situations where one really needs a reference to the search results generator, say to interleave calls to `next(search_results)` with other function calls, one can assign `search_results = client.search()` (without `await`) and then use the awaitable next method of the async generator.

```python
import asyncio

async def func():
    for i in range(10):
        yield i

async def main():
    gen = func()
    print("first")
    # In Python 3.10 this is: print(await anext(gen))
    print(await gen.__anext__())
    print("second")
    print(await gen.__anext__())

asyncio.run(main())
first
0
second
1
```

### PR DESCRIPTION
**Pull request by jreiberkyle**
**Related Issue(s):**

Closes #706 

Moves to wontfix: #476 and #779

**Proposed Changes:**

For inclusion in changelog (if applicable):

1. The following functions which handle paged results were changed to async generator functions and renamed:
  * OrdersClient.list_orders -> list_orders_aiter
  * DataClient.search -> search_aiter
  * DataClient.run_search -> run_search_aiter
  * DataClient.list_searches -> list_searches_aiter
3. The following were renamed to clarify that they are async generator functions:
  * SubscriptionsClient.list_subscription_results -> list_subscription_results_aiter
  * SubscriptionsClient.list_subscriptions -> list_subscriptions_aiter

Not intended for changelog:

1. 

**Diff of User Interface**

Old behavior:

e.g. for OrdersClient.list_orders:

```python
orders = await cl.list_orders()
orders_list = [o async for o in orders]
```

New behavior:

*note that cl.list_orders_aiter() is not awaited*

```python
orders_aiter = cl.list_orders_aiter()
orders_list = [o async for o in orders_aiter]
```

**PR Checklist:**

- [x] This PR is as small and focused as possible
- [x] If this PR includes proposed changes for inclusion in the changelog, the title of this PR summarizes those changes and is ready for inclusion in the Changelog.
- [x] I have updated docstrings for function changes and docs in the 'docs' folder for user interface / behavior changes
- [x] This PR does not break any examples or I have updated them

**(Optional) @mentions for Notifications:**
@JuLeeAtPlanet @cholmes

### CODE REVIEW

File: docs/python/sdk-guide.md [hunk 1]
```diff
@@ -109,13 +109,19 @@ Otherwise, the JSON blob is a list of the individual results.
...
+...         orders_list = collect(orders_aiter)
 ...
 >>> asyncio.run(main())
 
 ```
 
+Alternatively, these results can be converted to a list directly with
+
+```python
+>>>         orders_list = [o async for o in client.list_orders_aiter()]
```
[the remark this answers is not in the document]
Comment (kevinlacaille) [reply docs/python/sdk-guide.md:122]: Let's get rid of the large, confusing blank space.

**Comment by jreiberkyle** [comment]
@sgillies and @kevinlacaille - going ahead and merging this, mentioning you for an FYI

**Comment by kevinlacaille** [comment]
I'm not sure I see the value in adding `_aiter` to function & variable names. It seems like it only adds more confusion and verbosity. Especially since we're just appending `_aiter` to these names, it seems redundant. For instance, since there is no longer a `list_orders()` function why do we need to call it `list_orders_aiter()`?

**Review by kevinlacaille — COMMENTED** [review]
I'm a little uncertain about the `_aiter` suffix. See my comment above.

File: docs/python/sdk-guide.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jreiberkyle) [reply docs/python/sdk-guide.md:122]: yeah it's not great but I included it so the line could be plugged into the code example above. @JuLeeAtPlanet what are your thoughts on this?

**Comment by jreiberkyle** [comment]
@kevinlacaille I guess I can't reply to your review comment directly like in-line comments, which is unfortunate, but I have added an entry in the PR description that explains the reasoning behind the name change (see **Why the function name changes?** in the Proposed Changes section) . Thank you for digging deeper, it seems appropriate to provide more context around that change as it has a big impact.

File: docs/python/sdk-guide.md [hunk 2]
```diff
@@ -109,13 +109,19 @@ Otherwise, the JSON blob is a list of the individual results.
 >>> async def main():
 ...     async with Session() as sess:
 ...         client = OrdersClient(sess)
-...         orders = client.list_orders()
-...         orders_list = collect(orders)
+...         orders_aiter = client.list_orders_aiter()
+...         orders_list = collect(orders_aiter)
```
Comment (sgillies) [inline docs/python/sdk-guide.md:113]: @jreiberkyle @kevinlacaille let's not lead users to assign client.list_orders_aiter() to a variable. The best usage is just one single line.

```
orders = [order for order in client.list_orders()]
```

Lets recommend this and do it everywhere in this project.

I'm also :-1: on `_aiter`. There's no synchronous alternative and the fact that it's a generator is documented in the method docstring and in the type hint.

File: planet/cli/data.py [hunk 3]
```diff
@@ -272,12 +272,12 @@ async def search(ctx, item_types, filter, limit, name, sort, pretty):
...
-                                name=name,
-                                sort=sort,
-                                limit=limit)
-        async for item in items:
+        items_aiter = cl.search_aiter(item_types,
+                                      filter,
+                                      name=name,
+                                      sort=sort,
+                                      limit=limit)
+        async for item in items_aiter:
```
Comment (sgillies) [inline planet/cli/data.py:280]: Lines 275 could be a single line and improve clarity.

File: docs/python/sdk-guide.md (hunk 2 above)
Comment (sgillies) [reply docs/python/sdk-guide.md:113]: A note about the history of "iter" in names in In Python: whenever we've seen "iter" used as a prefix on method names, it's been part of a transition. Once, Python dicts had `keys()`, and it was a list. But then it was decided to make the keys method a generator function. To help the transition, `iterkeys()` was added. It returned a generator while `keys()` returned a list. And then in Python 3, `keys()` returned a generator and `iterkeys()` was deprecated.

**Pull request merged**